sql/src/main/java/io/crate/rest/action/RestResultSetReceiver.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.rest.action;
+
+import io.crate.action.sql.ResultReceiver;
+import io.crate.analyze.symbol.Field;
+import io.crate.concurrent.CompletionListener;
+import io.crate.concurrent.CompletionMultiListener;
+import io.crate.core.collections.Row;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.rest.BytesRestResponse;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestStatus;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+
+class RestResultSetReceiver implements ResultReceiver {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(RestResultSetReceiver.class);
+
+    private final List<Object[]> rows = new ArrayList<>();
+    private final RestChannel channel;
+    private final List<Field> outputFields;
+    private final boolean includeTypesOnResponse;
+    private long startTime;
+    private CompletionListener completionListener = CompletionListener.NO_OP;
+
+    RestResultSetReceiver(RestChannel channel,
+                          List<Field> outputFields,
+                          long startTime,
+                          boolean includeTypesOnResponse) {
+        this.channel = channel;
+        this.outputFields = outputFields;
+        this.includeTypesOnResponse = includeTypesOnResponse;
+        this.startTime = startTime;
+    }
+
+    @Override
+    public void addListener(CompletionListener listener) {
+        this.completionListener = CompletionMultiListener.merge(this.completionListener, listener);
+    }
+
+    @Override
+    public void setNextRow(Row row) {
+        rows.add(row.materialize());
```

Comment:
Wouldn't it be possible to write onto the XContentBuilder directly here? 
By doing that we could get rid of the `.materialize` calls